File: django_backend/authentication/services/redirect_service.py
from configurations.models import AuthConfiguration

DANGEROUS_SCHEMES = ["javascript:", "data:", "vbscript:", "file:"]

def validate_redirect_url(url_input):
    """
    Validate redirect URL.
    Returns (is_valid, sanitized_url_or_error_msg).
    """
    if not url_input or not isinstance(url_input, str):
        return False, "Please enter a valid HTTP(S) URL or relative path."
    
    clean_url = url_input.strip()
    lower_url = clean_url.lower()

    for scheme in DANGEROUS_SCHEMES:
        if lower_url.startswith(scheme):
            return False, "Please enter a valid HTTP(S) URL or relative path."

    if lower_url.startswith("http://") or lower_url.startswith("https://") or clean_url.startswith("/"):
        return True, clean_url

    # Relative path without leading slash
    if not ":" in clean_url:
        return True, clean_url

    return False, "Please enter a valid HTTP(S) URL or relative path."

def get_redirect_config(configuration_id=None, user_id=None):
    """
    Resolve configuration-specific redirect configuration dictionary.
    """
    config_obj = None

    if configuration_id:
        try:
            config_obj = AuthConfiguration.objects.filter(id=configuration_id, is_active=True).first()
        except Exception:
            pass

    if not config_obj and user_id:
        try:
            config_obj = AuthConfiguration.objects.filter(user_id=user_id, is_active=True).order_by("-updated_at").first()
        except Exception:
            pass

    default_redirect = {
        "enabled": True,
        "redirectUrl": "/dashboard",
        "redirectType": "url",
        "openInNewTab": False,
        "showSuccessMessage": True,
        "successMessage": "Authentication completed successfully.",
        "delay": 0
    }

    if not config_obj:
        return default_redirect

    data = config_obj.configuration_data or {}
    redirect_dict = data.get("redirect") if isinstance(data, dict) and isinstance(data.get("redirect"), dict) else {}

    raw_url = redirect_dict.get("redirectUrl") or config_obj.redirect_url or data.get("urls", {}).get("redirectUrl") or "/dashboard"
    is_valid, clean_url = validate_redirect_url(raw_url)
    if not is_valid:
        clean_url = "/dashboard"

    res = {
        "enabled": redirect_dict.get("enabled", True),
        "redirectUrl": clean_url,
        "redirectType": redirect_dict.get("redirectType", "url"),
        "openInNewTab": redirect_dict.get("openInNewTab", False),
        "showSuccessMessage": redirect_dict.get("showSuccessMessage", True),
        "successMessage": redirect_dict.get("successMessage", "Authentication completed successfully."),
        "delay": redirect_dict.get("delay", 0)
    }
    return res

def get_redirect_url(configuration_id=None, user_id=None):
    """
    Resolve configuration-specific redirect_url string.
    """
    config = get_redirect_config(configuration_id=configuration_id, user_id=user_id)
    return config["redirectUrl"] if config.get("enabled") else None

File: js/services/redirectService.js
/* =========================================================
   AUTH PAGE BUILDER - CENTRALIZED REDIRECT SERVICE
   File: js/services/redirectService.js
========================================================= */

(function (root, factory) {
  const service = factory();
  if (typeof module !== "undefined" && module.exports) {
    module.exports = service;
  }
  if (root) {
    root.RedirectService = service;
    root.redirectService = service;
  }
})(typeof window !== "undefined" ? window : globalThis, function () {

  let redirectExecuted = false;

  /**
   * Validate redirection URL.
   * Allows valid HTTP/HTTPS URLs and safe relative paths (e.g. /dashboard).
   * Rejects dangerous schemes: javascript:, data:, vbscript:, file:.
   */
  function validateUrl(urlInput) {
    if (urlInput === null || urlInput === undefined) {
      return { valid: false, error: "Please enter a valid HTTP(S) URL or relative path." };
    }

    const str = String(urlInput).trim();
    if (!str) {
      return { valid: false, error: "Please enter a valid HTTP(S) URL or relative path." };
    }

    const lower = str.toLowerCase();
    // Reject unsafe executable schemes explicitly
    const dangerousSchemes = ["javascript:", "data:", "vbscript:", "file:"];
    for (const scheme of dangerousSchemes) {
      if (lower.startsWith(scheme)) {
        return { valid: false, error: "Please enter a valid HTTP(S) URL or relative path." };
      }
    }

    // Check absolute HTTP/HTTPS or relative URL format
    if (lower.startsWith("http://") || lower.startsWith("https://")) {
      try {
        const parsed = new URL(str);
        if (parsed.protocol === "http:" || parsed.protocol === "https:") {
          return { valid: true, url: str };
        }
      } catch (_) {
        return { valid: false, error: "Please enter a valid HTTP(S) URL or relative path." };
      }
    }

    // Relative URL paths starting with / or alphanumeric relative paths
    if (str.startsWith("/") || /^[a-zA-Z0-9_\-\.\/]+$/.test(str)) {
      return { valid: true, url: str };
    }

    // Fallback URL parser attempt with relative base
    try {
      const origin = (typeof window !== "undefined" && window.location && window.location.origin) ? window.location.origin : "http://localhost";
      const resolved = new URL(str, origin);
      if (resolved.protocol === "http:" || resolved.protocol === "https:") {
        return { valid: true, url: str };
      }
    } catch (_) {}

    return { valid: false, error: "Please enter a valid HTTP(S) URL or relative path." };
  }

  /**
   * Resolve a relative or absolute URL into an absolute URL string for browser navigation.
   */
  function resolveTargetUrl(rawUrl) {
    if (!rawUrl) return "/dashboard";
    const str = String(rawUrl).trim();
    if (str.startsWith("http://") || str.startsWith("https://")) {
      return str;
    }
    try {
      const origin = (typeof window !== "undefined" && window.location && window.location.origin)
        ? window.location.origin
        : "http://localhost:3000";
      return new URL(str, origin).href;
    } catch (_) {
      return str;
    }
  }

  /**
   * Execute redirection.
   * THIS IS THE SOLE CHOKEPOINT FOR BROWSER NAVIGATION IN THE ENTIRE APPLICATION.
   *
   * @param {Object} redirectConfig - Canonical redirect config or partial override.
   * @param {Object} options - Options such as { isPreview: boolean, force: boolean, context: string }
   */
  function execute(redirectConfig = {}, options = {}) {
    if (redirectExecuted && options.force !== true && !options.isPreview) {
      console.log("[RedirectService] Redirection guard active. Duplicate execution prevented.");
      return Promise.resolve({ success: false, reason: "duplicate_prevented" });
    }

    // Normalize config values
    const config = {
      enabled: redirectConfig.enabled !== false,
      redirectUrl: redirectConfig.redirectUrl || redirectConfig.url || "/dashboard",
      redirectType: redirectConfig.redirectType || "url",
      openInNewTab: Boolean(redirectConfig.openInNewTab),
      showSuccessMessage: redirectConfig.showSuccessMessage !== false,
      successMessage: redirectConfig.successMessage || "Authentication completed successfully.",
      delay: Number(redirectConfig.delay) || 0
    };

    if (!redirectConfig || typeof redirectConfig !== "object") {
      console.log("[Redirect] No redirect configuration found");
      return Promise.resolve({ success: false, reason: "missing_config" });
    }

    if (!config.enabled) {
      console.log("[Redirect] Redirect disabled");
      return Promise.resolve({ success: false, reason: "disabled" });
    }

    // Validate URL scheme
    const valRes = validateUrl(config.redirectUrl);
    if (!valRes.valid) {
      const errToast = valRes.error || "Please enter a valid HTTP(S) URL or relative path.";
      if (typeof window !== "undefined" && window.Utils && typeof window.Utils.showToast === "function") {
        window.Utils.showToast(errToast, "error", 4000);
      }
      return Promise.reject(new Error(errToast));
    }

    // Activate single-execution guard
    redirectExecuted = true;

    const rawTarget = config.redirectUrl;
    const resolvedTarget = resolveTargetUrl(rawTarget);

    // Toast notification
    if (config.showSuccessMessage) {
      const displayMsg = `${config.successMessage} Redirect destination: ${rawTarget}`;
      const toastFn = (typeof window !== "undefined") ? (window.Utils?.showToast || window.showToast) : null;
      if (typeof toastFn === "function") {
        toastFn(displayMsg, "success", 3500);
      }
    }

    // Dispatch global redirect event for test runners / observers
    if (typeof window !== "undefined" && typeof window.dispatchEvent === "function") {
      try {
        window.dispatchEvent(
          new CustomEvent("auth:redirect", {
            detail: {
              url: rawTarget,
              resolvedUrl: resolvedTarget,
              config: config,
              isPreview: Boolean(options.isPreview)
            }
          })
        );
      } catch (e) {}
    }

    if (typeof window !== "undefined" && typeof window.onAuthRedirect === "function") {
      try {
        window.onAuthRedirect(rawTarget);
      } catch (e) {}
    }

    // Delay calculation
    const delayMs = Math.max(0, config.delay);

    return new Promise((resolve) => {
      setTimeout(() => {
        // If simulation explicitly requested by unit tests, suppress navigation
        if (options.simulateInPreview === true) {
          console.log("[Redirect] Simulated preview mode");
          console.log(`[Redirect] Target: ${rawTarget}`);
          resolve({ success: true, simulated: true, url: rawTarget, resolvedUrl: resolvedTarget });
          return;
        }

        // Sole Browser Navigation Execution Chokepoint
        try {
          if (rawTarget.startsWith("#")) {
            if (typeof window !== "undefined" && window.location) {
              window.location.hash = rawTarget;
            }
          } else if (config.openInNewTab) {
            if (typeof window !== "undefined" && typeof window.open === "function") {
              window.open(resolvedTarget, "_blank", "noopener,noreferrer");
            }
          } else {
            if (typeof window !== "undefined" && window.location && typeof window.location.assign === "function") {
              window.location.assign(resolvedTarget);
            } else if (typeof window !== "undefined" && window.location) {
              window.location.href = resolvedTarget;
            }
          }
        } catch (err) {
          console.error("[RedirectService] Navigation error:", err);
          if (typeof window !== "undefined" && window.location) {
            window.location.href = resolvedTarget;
          }
        }

        resolve({ success: true, url: rawTarget, resolvedUrl: resolvedTarget });
      }, delayMs);
    });
  }

  function resetGuard() {
    redirectExecuted = false;
  }

  function isExecuted() {
    return redirectExecuted;
  }

  return {
    validateUrl,
    resolveTargetUrl,
    execute,
    resetGuard,
    isExecuted
  };
});
